### ISSUE TITLE
The trajectory signal loop re-enters a per-position pass over policies that were just gathered as a batch

### ISSUE THREAD

**Issue by mmmfrieddough** [mmmfrieddough/anthro-chess#305]
## What was observed

Found while working #287, immediately downstream of it. #287 batched the
producer; this is the consumer it hands to.

**`_ScoringSession.trajectory` loops 610 times per batch over the three
policies `legal_policy_log_probabilities` just returned**, doing three
`policy_divergence` calls, two `argmax` calls and four `.item()` reads per
position:

```python
for offset, key in enumerate(zip(active.game_ids, active.ply_indices, strict=True)):
    signals[key] = _trajectory_signal(
        legal_actions=self._inputs.plies[key].enabled_actions(),
        target_action_id=self._inputs.plies[key].target_action_id,
        true=true[offset], low=low[offset], high=high[offset],
    )
```

Measured on this host at the real evaluation batch shape — 610 enabled rows,
vocabulary 1970 — this loop costs **13.9 ms per batch**, against **9.1 ms** for
the three `legal_policy_log_probabilities` calls that feed it. After #287 it is
the larger of the two, and the largest remaining per-position host cost in the
dependency pass.

Every quantity it computes is a reduction over the same segments #287 gathers
as one block. `policy_divergence` is a sum over one position's legal entries;
`argmax` is a per-segment maximum. Both have batched forms over the flat
gathered block plus a row index — which #287 does not carry, having settled on
gathering through `ActiveBatch.legal_mask` instead, so whoever takes this
either derives the row vector once per batch (`repeat_interleave` over the row
lengths, about 54 us at this shape) or revisits that decision with this
consumer's needs on the table. The four `.item()` reads per position are
float pulls that one `.tolist()` per quantity would replace.

The one part that is genuinely per-position is `legal_actions.index(target_action_id)`
in `_trajectory_signal`, a linear scan to find the target's offset within its
row.

## Why

Same shape as #211, #284 and #287: per-position host work in front of a tensor
operation, on the command whose GPU sits under 1% busy. This one is what #287
left as the majority term.

## What it buys

Zero change to any reading, if the batched reductions reproduce the current
ordering and float64 accumulation.

## What would settle it

Compute the three divergences and the two argmaxes as segment reductions over
the gathered block, read them back with one `.tolist()` per quantity, and build
the `TrajectorySignal` records from the resulting lists. Keep float64.

## Measurement context

One Linux host, 2x RTX 4090, 32 cores, CUDA, both GPUs idle. Per-call figures
at tensor shapes taken from real batches over `lichess-blitz-2017-04-test`.

## References

- #287 — batched the producer this loop consumes, and measured this loop as the
  larger remaining term
- #284, #211 — where the sequence started

**Comment by mmmfrieddough** [comment]
Measured while working #306, since #306's normalize needed the same row vector this issue wants.

**Carrying a flat `row * ACTION_VOCABULARY_SIZE + column` index on `ActiveBatch` is worth about 36% of `legal_policy_log_probabilities`, with bitwise identical output.** At the real evaluation batch shape — 610 enabled rows, vocabulary 1970, 18,643 legal entries:

| variant | per call | per batch (3 calls) |
| --- | ---: | ---: |
| deriving the row vector locally (what #306 landed) | 0.589 ms | 1.77 ms |
| flat index carried on `ActiveBatch` | **0.378 ms** | **1.13 ms** |

The saving is not the row vector, which is cheap. It is that `logits.flatten()[index]` does not have to scan all 1,201,700 boolean mask entries to find the legal ones, the way `logits[legal_mask]` does — that scan is 0.148 ms of the 0.589 ms.

I did not take it in #306. One call site did not justify a field on a frozen value type every scorer shares, threaded through `rescored()` and needing test scaffolding because dataclass equality cannot compare tensors — and #287 declined the same change at a 0.137 ms/call win. This issue is the second consumer, so it is where the field pays for itself.

Two other figures from the same measurements that bear on this issue:

- `split_with_sizes` is now the largest single step in the producer at **0.191 ms**, 25–31% of the call. It exists only to hand back one tensor per position. This issue's batched form wants the flat block instead, so it removes that term rather than optimizing it.
- `_legal_mask` already builds both the row and column index vectors and discards them, so the index above is a retention rather than a new computation. It costs `_legal_mask` about 0.04 ms once per batch to keep.

Measured on one Linux host, 32 cores, 2x RTX 4090 both idle, CPU float64, at tensor shapes taken from real batches over `lichess-blitz-2017-04-test`.

**Comment by mmmfrieddough** [comment]
Correcting my own figures above — they came from a fresh-process benchmark and are inflated.

Measured **in situ**, inside a real `anthro eval run` at the same shape (410 enabled rows, 13,434 legal entries):

| gather | in situ | my earlier fresh-process figure |
| --- | ---: | ---: |
| `logits[legal_mask]` | 0.0754 ms | 0.148 ms |
| `logits.flatten()[index]` | 0.0123 ms | 0.021 ms |

So carrying the index saves about **0.063 ms per call**, not the 0.21 ms implied above — roughly **3.6 ms across a 128-game view**, against a command that takes ~25 s. This part is a mask *scan* rather than an allocation, so it does survive in a warm process, unlike the normalize saving; it is just much smaller than a fresh-process benchmark suggests.

**The general caution, which cost me real time here:** a fresh Python process pays mmap/munmap and page-fault costs on every multi-megabyte tensor allocation that the real evaluation process does not, because that process holds a large working heap from the model and loader. The same captured `ActiveBatch` and the same code measured **2.16 ms per call in a clean harness and 0.52 ms inside the eval process**. Anything in this issue family should be timed in situ before the win is believed.

That applies to this issue's own headline as well: the 13.9 ms per batch for the trajectory loop in the body is a fresh-process figure. The loop is mostly Python-level work rather than large allocations, so it will hold up better than the normalize did — but it is worth re-measuring in situ before sizing the work, because the whole trajectory pass is only ~1.19 s per 128-game view and that bounds what this can win.

**Comment by mmmfrieddough** [comment]
Parked. 13.9 ms per batch of host work, and the issue states it makes zero change to any reading. 

Routing, not rejection, per the bar in `docs/issue-workflow.md`: it improves something that already produces a correct reading, so it leaves the queue sessions select from next and keeps its evidence. It returns by being moved into a numbered milestone deliberately.
